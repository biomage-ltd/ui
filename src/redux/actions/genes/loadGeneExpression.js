import {
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_ERROR, GENES_EXPRESSION_LOADED,
} from '../../actionTypes/genes';

import { fetchCachedWork } from '../../../utils/cacheRequest';

const loadGeneExpression = (
  experimentId, genes, componentUuid, forceReloadAll = false,
) => async (dispatch, getState) => {
  const {
    loading, data: geneData,
  } = getState().genes.expression;

  // If other gene expression data is already being loaded, don't dispatch.
  if (loading.length > 0) {
    return null;
  }

  console.log('genes updating with ', genes);

  // Dispatch loading state.
  dispatch({
    type: GENES_EXPRESSION_LOADING,
    payload: {
      experimentId,
      genes,
      componentUuid,
    },
  });

  // Check which of the genes we actually need to load. Only do this if
  // we are not forced to reload all of the data.
  let genesToFetch = [...genes];

  if (!forceReloadAll) {
    const genesAlreadyLoaded = new Set(Object.keys(geneData));
    genesToFetch = genesToFetch.filter((gene) => !genesAlreadyLoaded.has(gene));
  }

  if (genesToFetch.length === 0) {
    return dispatch({
      type: GENES_EXPRESSION_LOADED,
      payload: {
        experimentId,
        componentUuid,
        genes,
      },
    });
  }

  const body = {
    name: 'GeneExpression',
    genes: genesToFetch,
  };

  try {
    const data = await fetchCachedWork(experimentId, 30, body);
    if (Object.keys(data).length === 0) {
      throw Error('There is no information available for selected genes.');
    }
    dispatch({
      type: GENES_EXPRESSION_LOADED,
      payload: {
        experimentId,
        data,
        componentUuid,
        genes,
      },
    });
  } catch (e) {
    dispatch({
      type: GENES_EXPRESSION_ERROR,
      payload: {
        experimentId,
        error: "Couldn't fetch expression data.",
        componentUuid,
      },
    });
  }
};

export default loadGeneExpression;
