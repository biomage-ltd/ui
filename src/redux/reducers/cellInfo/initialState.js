const initialState = {
  focus: {
    store: 'cellSets',
    key: 'louvain',
  },
  groupedTrack: 'louvain',
  selectedTracks: ['louvain'],
};

export default initialState;
