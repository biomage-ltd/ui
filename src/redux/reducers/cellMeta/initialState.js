const mitochondrialContent = {
  loading: true,
  error: false,
  data: [],
};

const doubletScores = {
  loading: true,
  error: false,
  data: [],
};

const initialState = {
  mitochondrialContent,
  doubletScores,
};

export default initialState;
