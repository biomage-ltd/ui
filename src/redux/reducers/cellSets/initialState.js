const initialState = {
  properties: {},
  hierarchy: [],
  selected: [],
  loading: true,
  error: false,
  updatingClustering: false,
  hidden: new Set(),
};

export default initialState;
