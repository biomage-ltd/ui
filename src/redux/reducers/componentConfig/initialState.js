import {
  legendBaseState,
  dimensionsBaseState,
  axesBaseState,
  titleBaseState,
  fontStyleBaseState,
  colourBaseState,
  markerBaseState,
  labelBaseState,
} from './baseStylesState';

// PLOTS & TABLES - Categorical Embedding
const embeddingCategoricalInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// PLOTS & TABLES - Continuous Embedding
const embeddingContinuousInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    dx: 0,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  logEquation: 'datum.expression*1',
  shownGene: 'notSelected',
  selectedSample: 'All',
};

// PLOTS & TABLES - Heatmap
const heatmapInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    show: true,
    position: 'horizontal',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 500,
    height: 500,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
    dx: 0,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedGenes: [],
  selectedCellSet: 'louvain',
  labelColour: 'transparent',
};

// PLOTS & TABLES - Volcano plot
const volcanoInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: dimensionsBaseState,
  marker: {
    ...markerBaseState,
    size: 32,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log2 fold change',
    yAxisText: ' - log10(adj.p - value)',
    gridOpacity: 5,
    offset: 10,
  },
  title: titleBaseState,
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  label: labelBaseState,
  noDifferenceColor: '#aaaaaa',
  significantUpregulatedColor: '#0000ffaa',
  significantDownregulatedColor: '#ff0000',
  notSignificantDownregulatedColor: '#aaaaaa',
  notSignificantUpregulatedColor: '#aaaaaa',
  significantChangeDirectionUnknownColor: '#aaaaaa',

  // `null` automatically scales to range. This is a problem
  // because our DE is bad right now, so it throws off the
  // range to extreme values. TODO: set this back when we have
  // good DE
  // logFoldChangeDomain: null,

  logFoldChangeDomain: 20,
  maxNegativeLogpValueDomain: null,
  negLogpValueThreshold: 4,
  logFoldChangeThreshold: 1,
  logFoldChangeTickCount: 5,
  negativeLogpValueTickCount: 5,
  downsampleRatio: 0,
  showLogFoldChangeThresholdGuides: true,
  showpvalueThresholdGuides: true,
  thresholdGuideWidth: 1,
  logFoldChangeThresholdColor: '#ff0000',
  pvalueThresholdColor: '#ff0000',
  textThresholdValue: 240,
  strokeOpa: 1,
  strokeCol: '#000000',
};

// PLOTS & TABLES - Frequency
const frequencyInitialConfig = {
  spec: '1.0.0',
  frequencyType: 'proportional',
  proportionGrouping: '',
  legend: {
    ...legendBaseState,
    title: 'Cell Set',
    position: 'top',
    offset: 40,
  },
  label: labelBaseState,
  dimensions: dimensionsBaseState,
  marker: markerBaseState,
  colour: colourBaseState,
  title: titleBaseState,
  axes: {
    ...axesBaseState,
    xAxisText: 'Sample',
    yAxisText: 'Proportion',
    offset: 10,
  },
  fontStyle: fontStyleBaseState,
  xAxisGrouping: '',
  axisTitlesize: 13,
  geneexpLegendloc: '',
};

// EMBEDDING PREVIEW - Coloured by sample
const embeddingPreviewBySampleInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// EMBEDDING PREVIEW - Coloured by cell sets
const embeddingPreviewByCellSetsInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// EMBEDDING PREVIEW - Config for fraction of mitochondrial reads
const embeddingPreviewMitochondrialContentInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    dx: 0,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedSample: 'All',
};

// EMBEDDING PREVIEW - Config for doublet score
const embeddingPreviewDoubletScoreInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    dx: 0,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedSample: 'All',
};

const interactiveHeatmapInitialConfig = {
  selectedTracks: ['louvain'],
  groupedTracks: ['sample', 'louvain'],
  expressionValue: 'raw',
  legendIsVisible: true,
};

// CELL SIZE DISTRIBUTION - Cell Size Distribution Histogram
const cellSizeDistributionHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-left',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: '#UMIs in cell',
    yAxisText: '#UMIs * #Cells',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  minCellSize: 10800,
  binStep: 200,
};

// CELL SIZE DISTRIBUTION - Cell Size Distribution Knee Plot
const cellSizeDistributionKneePlot = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-left',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'Cell rank',
    yAxisText: '#UMIs in cell',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  minCellSize: 990,
};

// MITOCHONDRIAL CONTENT - Mitochondrial Fraction Histogram
const mitochondrialFractionHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-right',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'Fraction of mitochondrial reads',
    yAxisText: 'Fraction of cells',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
    dx: 0,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  binStep: 0.05,
  maxFraction: 0.1,
};

// MITOCHONDRIAL CONTENT - Mitochondrial Fraction Log Histogram
const mitochondrialFractionLogHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-right',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10(#UMIs in cell)',
    yAxisText: 'Fraction of mitochondrial reads',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
    dx: 0,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  binStep: 0.05,
  maxFraction: 3.5,
};

// CLASSIFIER - Classifier Empty Drops
const classifierEmptyDropsPlot = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    enabled: false,
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 630,
    height: 500,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10[ cell size (UMIs) ]',
    yAxisText: 'classifier prob',
    gridOpacity: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  minProbability: 0.82,
  bandwidth: -1,
};

// GENES VS UMIS - Features vs UMIs
const featuresVsUMIsHistogram = {
  spec: '1.0.0',
  dimensions: {
    ...dimensionsBaseState,
    width: 630,
    height: 500,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10 [molecules]',
    yAxisText: 'Frequency',
    gridOpacity: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  label: labelBaseState,
  lowCutoff: 4.8,
  upCutoff: 2.1,
  binStep: 0.05,
};

// GENES VS UMIS - Features vs UMIs
const featuresVsUMIsScatterplot = {
  spec: '1.0.0',
  dimensions: {
    ...dimensionsBaseState,
    width: 630,
    height: 500,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10 [molecule counts]',
    yAxisText: 'log10 [gene counts]',
    gridOpacity: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  label: labelBaseState,
  lowCutoff: 4.8,
  upCutoff: 2.1,
};

// DOUBLE FILTER - Doublet filter histogram
const doubletScoreHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-right',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'Probability of being a doublet',
    yAxisText: 'Frequency',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  binStep: 0.05,
  probThreshold: 0.2,
};

// DATA INTEGRATION - Embedding by Samples
const dataIntegrationEmbeddingInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// DATA INTEGRATION - Frequency
const dataIntegrationFrequencyInitialConfig = {
  ...frequencyInitialConfig,
  xAxisGrouping: 'louvain',
  proportionGrouping: 'sample',
  legend: {
    ...legendBaseState,
    title: 'Sample',
    position: 'top',
    offset: 10,
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },

  axes: {
    ...axesBaseState,
    xAxisText: 'Louvain clusters',
    yAxisText: 'Proportion',
    offset: 10,
  },
  fontStyle: fontStyleBaseState,
};

// DATA INTEGRATION - Elbow
const dataIntegrationElbowPlotInitialConfig = {
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  label: { ...labelBaseState },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  marker: { ...markerBaseState },
  fontStyle: fontStyleBaseState,
  axes: {
    ...axesBaseState,
    xAxisText: 'Principal Components',
    yAxisText: 'Proportion of Variance Explained',
    titleFont: 'sans-serif',
    labelFont: 'sans-serif',
    labelFontSize: 13,
    offset: 0,
    gridOpacity: 10,
  },
  colour: {
    toggleInvert: '#FFFFFF',
  },
  title: {
    ...titleBaseState,
    font: 'sans-serif',
    fontSize: 12,
  },
  signals: [
    {
      name: 'interpolate',
      value: 'linear',
      bind: {
        input: 'select',
        options: [
          'basis',
          'cardinal',
          'catmull-rom',
          'linear',
          'monotone',
          'natural',
          'step',
          'step-after',
          'step-before',
        ],
      },
    },
  ],
};

const initialPlotConfigStates = {
  cellSizeDistributionHistogram,
  cellSizeDistributionKneePlot,
  mitochondrialFractionHistogram,
  mitochondrialFractionLogHistogram,
  classifierEmptyDropsPlot,
  featuresVsUMIsHistogram,
  featuresVsUMIsScatterplot,
  doubletScoreHistogram,
  embeddingCategorical: embeddingCategoricalInitialConfig,
  embeddingContinuous: embeddingContinuousInitialConfig,
  heatmap: heatmapInitialConfig,
  volcano: volcanoInitialConfig,
  frequency: frequencyInitialConfig,
  embeddingPreviewBySample: embeddingPreviewBySampleInitialConfig,
  embeddingPreviewByCellSets: embeddingPreviewByCellSetsInitialConfig,
  embeddingPreviewMitochondrialContent: embeddingPreviewMitochondrialContentInitialConfig,
  embeddingPreviewDoubletScore: embeddingPreviewDoubletScoreInitialConfig,
  dataIntegrationEmbedding: dataIntegrationEmbeddingInitialConfig,
  dataIntegrationFrequency: dataIntegrationFrequencyInitialConfig,
  dataIntegrationElbow: dataIntegrationElbowPlotInitialConfig,
};

const initialComponentConfigStates = {
  interactiveHeatmap: interactiveHeatmapInitialConfig,
};
export { initialPlotConfigStates, initialComponentConfigStates };

const initialState = {};
export default initialState;
