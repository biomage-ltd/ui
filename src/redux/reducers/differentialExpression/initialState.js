const initialState = {
  properties: {
    data: [],
    cellSets: {},
    loading: false,
    error: false,
    total: 0,
  },
  comparison: {
    type: 'between',
    group: {
      cellSet: null,
      compareWith: null,
      basis: null,
    },
  },
};

export default initialState;
