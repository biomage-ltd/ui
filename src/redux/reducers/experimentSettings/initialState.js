const initialPipelineState = {
  startDate: null,
  endDate: null,
  status: null,
  completedSteps: [],
};

const initialWorkerState = {
  status: null,
  started: false,
  ready: null,
  restartCount: 0,
};

const initialState = {
  pipelineStatus: {
    loading: false,
    error: false,
    status: {
    },
  },
  info: {
    experimentId: null,
    experimentName: null,
    projectId: null,
    createdAt: null,
    lastViewed: null,
  },
  processing: {
    meta: {
      loading: true,
      completingStepError: false,
      loadingSettingsError: false,
      selectedConfigureEmbeddingPlot: 'sample',
    },
  },
};

export default initialState;
export { initialPipelineState, initialWorkerState };
