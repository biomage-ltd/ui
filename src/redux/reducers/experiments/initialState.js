const experimentTemplate = {
  projectUuid: null,
  name: null,
  description: 'Analysis description',
  id: null,
  createdDate: null,
  lastViewed: null,
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
  },
  ids: [],
};

export default initialState;
export { experimentTemplate };
