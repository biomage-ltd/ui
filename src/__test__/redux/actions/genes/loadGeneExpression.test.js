import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import loadGeneExpression from '../../../../redux/actions/genes/loadGeneExpression';
import initialState from '../../../../redux/reducers/genes/initialState';
import sendWork from '../../../../utils/sendWork';

import {
  GENES_EXPRESSION_LOADING,
  GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
} from '../../../../redux/actionTypes/genes';

jest.mock('localforage');
jest.mock('../../../../utils/sendWork', () => ({
  __esModule: true, // this property makes it work
  default: jest.fn(),
}));

const mockStore = configureStore([thunk]);

describe('loadGeneExpression action', () => {
  const experimentId = '1234';
  const componentUuid = 'asd';
  const loadingGenes = ['a', 'b', 'c'];

  it('Does not dispatch when expression is already loading', async () => {
    const store = mockStore({
      genes:
        { ...initialState, expression: { ...initialState.expression, loading: ['d'] } },
    });

    store.dispatch(loadGeneExpression(experimentId, loadingGenes, componentUuid));
    expect(store.getActions().length).toEqual(0);
  });

  it('Does not send work for already loaded expression data.', async () => {
    const store = mockStore({
      genes: {
        ...initialState,
        expression: {
          ...initialState.expression,
          data: {
            ...initialState.expression.data,
            a: {
              min: 0,
              max: 0,
              expression: [0, 0, 0, 0, 0],
            },
          },
        },
      },
    });

    sendWork.mockImplementation(() => {
      // No need to mock the result accurately.

      const resolveWith = {
        results:
          [
            {},
          ],
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    await store.dispatch(loadGeneExpression(experimentId, loadingGenes, componentUuid));

    const firstCall = sendWork.mock.calls[0];
    expect(firstCall[2].genes).toEqual(['b', 'c']);
    expect(sendWork).toMatchSnapshot();
  });

  it('Sends work for already loaded expression data if forced to do so.', async () => {
    const store = mockStore({
      genes: {
        ...initialState,
        expression: {
          ...initialState.expression,
          data: {
            ...initialState.expression.data,
            a: {
              min: 0,
              max: 0,
              expression: [0, 0, 0, 0, 0],
            },
          },
        },
      },
    });

    sendWork.mockImplementation(() => {
      // No need to mock the result accurately.

      const resolveWith = {
        results:
          [
            {},
          ],
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    await store.dispatch(loadGeneExpression(experimentId, loadingGenes, componentUuid, true));

    const firstCall = sendWork.mock.calls[1];
    expect(firstCall[2].genes).toEqual(['a', 'b', 'c']);
    expect(sendWork).toMatchSnapshot();
  });

  it('Dispatches appropriately on success', async () => {
    const store = mockStore({
      genes: {
        ...initialState,
      },
    });

    // Need to mock result accurately because of post-request processing
    const mockResolve = {
      geneA: {
        expression: [1],
        mean: 1,
        stdev: 1,
      },
    };

    const mockResult = {
      geneA: {
        expression: [1],
        mean: 1,
        stdev: 1,
        zScore: [0],
      },
    };

    sendWork.mockImplementation(() => {
      const resolveWith = {
        results:
          [
            {
              body: JSON.stringify(mockResolve),
            },
          ],
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    await store.dispatch(loadGeneExpression(experimentId, loadingGenes, componentUuid, true));

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(GENES_EXPRESSION_LOADING);
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(GENES_EXPRESSION_LOADED);
    expect(loadedAction.payload.data).toEqual(mockResult);
    expect(loadedAction).toMatchSnapshot();
  });

  it('Dispatches appropriately on failure', async () => {
    const store = mockStore({
      genes: {
        ...initialState,
      },
    });

    sendWork.mockImplementation(() => new Promise((resolve, reject) => reject(new Error('random error!'))));
    await store.dispatch(loadGeneExpression(experimentId, loadingGenes, componentUuid, true));

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(GENES_EXPRESSION_LOADING);
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(GENES_EXPRESSION_ERROR);
    expect(loadedAction).toMatchSnapshot();
  });
});
