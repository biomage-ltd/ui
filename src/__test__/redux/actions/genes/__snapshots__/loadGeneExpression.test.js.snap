// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadGeneExpression action Dispatches appropriately on failure 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
    "genes": Array [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/expressionLoading",
}
`;

exports[`loadGeneExpression action Dispatches appropriately on failure 2`] = `
Object {
  "payload": Object {
    "error": "Couldn't fetch expression data.",
    "experimentId": "1234",
  },
  "type": "genes/expressionError",
}
`;

exports[`loadGeneExpression action Dispatches appropriately on success 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
    "genes": Array [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/expressionLoading",
}
`;

exports[`loadGeneExpression action Dispatches appropriately on success 2`] = `
Object {
  "payload": Object {
    "data": Object {
      "a": "b",
    },
    "experimentId": "1234",
  },
  "type": "genes/expressionLoaded",
}
`;

exports[`loadGeneExpression action Does not send work for already loaded expression data. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "1234",
      30,
      Object {
        "genes": Array [
          "b",
          "c",
        ],
        "name": "GeneExpression",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`loadGeneExpression action Sends work for already loaded expression data if forced to do so. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "1234",
      30,
      Object {
        "genes": Array [
          "b",
          "c",
        ],
        "name": "GeneExpression",
      },
    ],
    Array [
      "1234",
      30,
      Object {
        "genes": Array [
          "a",
          "b",
          "c",
        ],
        "name": "GeneExpression",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
